### DISCUSSION TITLE
Behavior on overflow (`Index`, `Signed`, ....)

### DISCUSSION THREAD

**Discussion by martijnbastiaan** [clash-lang/clash-compiler#1903] [Ideas]
#688 has been open for almost two years now, and I feel bad that we haven't responded to it properly. This mostly stems from the fact that its a "controversial" idea, with lots of different opinions. I'm opening a discussion for two reasons:

* To reach a conclusion on that PR, even if it ends in closing/rejecting it.
* To test GitHub discussions :)

The gist of this issue is captured in this trace:

```haskell
>>> (3 :: Index 4) + 1
<error>
>>> (3 :: Unsigned 2) + 1
0
```

There are a couple of hypothetical reasons this decrepancy is currently there:

* In most programming languages `Unsigned` and `Signed` do not raise an error on overflow. Either because their implementation is bigint (Python, Ruby) or because of -presumably- performance considerations (Haskell, C++, Java). A notable exception to this rule is :crab: Rust :crab:, it throws an exception in "debug" mode and reverts to silently overflowing on "release" mode.  
* Clash inherits Haskell's overflow behavior for `Int`, it makes sense to mirror them for `Unsigned`/`Signed`/..
* `Index` needs bound checking anyway to be useful, so raising an error is not a performance consideration.

# Options
I believe we have a couple of options. If you'd like one added to the list, please edit this post or leave a comment.

## Option 1: keep as is (reject PR)
Self-explanatory.

## Option 2: make saturation mode wrapping by default for `Index` (~accept PR)
Do not error on overflow for operations on `Index`. (I'd like to save discussions on an exact implemenation for later.)

## Option 3: make saturation mode non-wrapping (error) by default for `Unsigned`, `Signed`, ..
Similar to Rust, we could make overflowing an error for _all_ Clash data types. 

# Type-level saturation mode
PR #688 also implements a way of specifying saturation mode at the type-level. However, I think we pretty much all agree that @alex-mckenna's [proposal](https://github.com/clash-lang/clash-compiler/pull/688#issuecomment-835825071) makes most sense here. If this does happen to be contentious, we should discuss it elsewhere.

**Comment by martijnbastiaan** [comment]
My personal opinion on this matter can be captured in a few points:

* I strongly believe that silent overflowing is an historical coincidence, of which the biggest part was "ease of implementation" and in a smaller part "performance". I believe this due to C (and perhaps C++) defining overflow on signed integers as _undefined_. 
* I believe that in most situations silent overflow is not what the developer intented. This argument is weakened by the fact that Clash developers choose their integer widths, but my guess is it still holds.
* I believe the following points are relevant in this discussion (from the Python Zen, ofc):
  * Explicit is better than implicit.
  * In the face of ambiguity, refuse the temptation to guess.
  * Errors should never pass silently.

These points lead me to believe the orange crab language got it right once more, and we should go with _Option 3: make saturation mode non-wrapping (error) by default for Unsigned, Signed, .._.

----------------

Couple of points:

* I wouldn't be opposed to a `-fno-oob-checks` on `clash-prelude` that'd model the debug/release mode of Rust.
* `Int` and friends would remain default Haskell behavior, but I think usage of those types should be discouraged in Clash anyway.

**Reply by alex-mckenna** [reply]
> C (and perhaps C++) defining overflow on unsigned integers as undefined

FWIW C/C++ define overflow on signed numbers as being UB (although most implementations wrap, they don't have to) but overflow on unsigned is impossible because it's defined as always wrapping. To quote the C standard:

> A computation involving unsigned operands can never overflow, because a result that cannot be represented by the resulting unsigned integer type is reduced modulo the number that is one greater than the largest value that can be represented by the resulting type.

**Reply by alex-mckenna** [reply]
I would also be somewhat against a flag in `clash-prelude` for changing the default. It's not "typical" Haskell to keep changing package flags during development like this IME and I think it would end up being ignored by most users

**Reply by martijnbastiaan** [reply]
Woops, I  meant _signed_.

**Comment by christiaanb** [comment]
Clash wraps around on overflow (for `Signed` and `Unsigned`) because that's how numerical operators (adders, multipliers, etc.) in their most basic form work in hardware, so it's not an historical coincidence.

**Reply by martijnbastiaan** [reply]
> because that's how numerical operators (adders, multipliers, etc.) in their most basic form work in hardware, so it's not an historical coincidence.

The industry happened to converge on 2s-complement, making wrap-around a convenient overflow method for `Signed`. This is reflected in the C spec - it explicitely makes `Signed` wraparound as undefined behavior. Similarly, wrap-around on `Unsigned` was defined to be wrap-around because pretty much all hardware could implement it efficiently.  And this is exactly what I meant by "historical coincidence": there's no overarching design philosophy. There's just "ease of implementation" and for a smaller part "performance". If it had turned out that `Signed` arithmetic would be easier with overflow-to-zero, that would have been the standard today.

**Reply by christiaanb** [reply]
Performance is the major argument for doing wrap-around on overflow, that is, I wanted the basic numeric operators to map as closely to their (by these days) standard hardware implementation. And you can build the more "expensive" (in terms of performance) overflow behavior on top of it: 
- Error out: mux with undefined, logic synthesis has the option to optimize that mux away
- Saturation: mux with max, mux will not be optimized away.

**Reply by christiaanb** [reply]
And the same performance argument holds for wrap-around and 2s complement interpretation: those operators are the smallest to implement in cmos.

**Reply by martijnbastiaan** [reply]
> Error out: mux with undefined, logic synthesis has the option to optimize that mux away

Which it will, so erroring on overflow (in simulation) will not affect the generated hardware. Clash can even remove these branches itself with `-faggressive-x-optimization` AFAIK, which is not an expensive transformation at all. I'm not convinced you'll be able to measure the difference - for synthesis tools or Clash. What remains is simulation speed.

Could you weigh this against the other design points I posted though? We've known for some time now that dodgy overflow behavior has lead to disastrous bugs and we see newer languages either cut ties with it using compiler flags (Rust) or switching to bigint (Python, Ruby).

**Reply by martijnbastiaan** [reply]
Just to be clear: I'm not advocating for the removal of wrap-around behavior. I'm advocating for the removal of the _default_ being wrap-around on grounds of:

 * Explicit is better than implicit.
 * In the face of ambiguity, refuse the temptation to guess.
 * Errors should never pass silently.

Access to wrap-around behavior should remain easy either through newtype wrappers or `satAdd` and friends.

**Reply by rowanG077** [reply]
I agree with @martijnbastiaan. If Clash can ensure the undefined branches are pruned during synthesis then the only real argument for me would be simulation performance. 

But it's important how backwards compatibility is handled. This would be an API change, is it possible to give a deprecation notice instead of an error when overflow occurs during simulation? That way we could write something like: warning overflow detected. Please use this and that to get proper overflow behavior. this warning will turn into an error from clash x.y.z forward.

**Reply by christiaanb** [reply]
But when error'ing out is the default, how do you get zero-cost wrap-around as a user? What's the proposed API there? With wrap-around as the default, we can add error-out as a zero-cost abstraction on top.

**Reply by rowanG077** [reply]
The api @alex-mckenna proposed. The default `Num` instance would error out in simulation similar to what `Index` does now. If a users wants to have wrap-around behaviour they need to wrap their numerical type in the `Wrapping` newtype. See here: https://github.com/clash-lang/clash-compiler/commit/c17ce246821a5c5afca43bea88568743bc8cd59f

**Reply by martijnbastiaan** [reply]
If we'd merge @alex-mckenna's proposal we'd have ~~to~~ two options I think:

* ```a `wrapAdd` b``` (and friends). This would be similar to Rust's [wrapping_add](https://doc.rust-lang.org/std/intrinsics/fn.wrapping_add.html).
* ``` a + b``` in combination with the newtype wrappers

**Reply by christiaanb** [reply]
Right, so primitives like `Clash.Sized.Internal.Unsigned.+# :: Unsigned n -> Unsigned n -> Unsigned n` will still wrap around on overflow.

**Reply by martijnbastiaan** [reply]
Oh yeah, internal primitives would remain the same. I was just thinking about the API we expose to users.

**Reply by martijnbastiaan** [reply]
(Sorry didn't see your comment @rowanG077)

**Reply by martijnbastiaan** [reply]
> But it's important how backwards compatibility is handled. This would be an API change, is it possible to give a deprecation notice instead of an error when overflow occurs during simulation? That way we could write something like: warning overflow detected. Please use this and that to get proper overflow behavior. this warning will turn into an error from clash x.y.z forward.

I agree, we should think about this and that's probably possible. However, I'd like to postpone discussing implementation details a bit and focus on the direction we're heading. If an option turns out to be infeasible to implement, we can revisit. I don't want another two years of limbo :smile: .

**Comment by alex-mckenna** [comment]
For reference: my proposal also comes with a mostly finished implementation at [`clash-num-newtypes`](https://github.com/clash-lang/clash-compiler/tree/clash-num-newtypes)

**Reply by martijnbastiaan** [reply]
This is great :). Any particular reason it's stuck, or just some loose ends?

**Reply by alex-mckenna** [reply]
IIRC the tests were fiddly to get right and then I got distracted by something else. _So it goes_

**Reply by martijnbastiaan** [reply]
So it goes.

**Comment by gergoerdi** [comment]
So `x + negate x == 0` wouldn't be true anymore for `Unsigned`?

**Reply by basile-henry** [reply]
Yeah I think it would error out with this proposal. But `x + negate x == 0` would be true for `Wrapping (Unsigned n)` with Alex' suggestions :smile:

**Reply by alex-mckenna** [reply]
It would still be true with option 2, I guess @martijnbastiaan prefers option 3 though. I don't think we have consensus on any particular direction yet though, at least not that I'm aware

**Reply by martijnbastiaan** [reply]
Either @basile-henry's suggestion or ```x `wrapAdd` negate x == 0``` would work. 

> It would still be true with option 2, I guess @martijnbastiaan prefers option 3 though

Yes, I'm in favor of option 3. I'd like to hear other POVs though!

**Reply by basile-henry** [reply]
> or x `wrapAdd` negate x == 0 would work

I thought `negate` would still be an issue here if `Unsigned` is not allowed wrap.

**Reply by alex-mckenna** [reply]
Yes, it would error on every call to `negate x` where `x /= 0`. So you'd also want to use a safe `negate` like the one in `Wrapping`

**Reply by martijnbastiaan** [reply]
> I thought negate would still be an issue here if Unsigned is not allowed wrap.

Yep. I think we need a :facepalm: reaction emoji.

**Comment by gergoerdi** [comment]
Completely anecdotal data point: I have hacked Clash 1.4 locally so that `+#` and `-#` error out on over/underflows, to simulate option #3, and then tried to get https://github.com/gergoerdi/clash-compucolor2 working again. It... wasn't pleasant at all.

Even though it's quite a small code base, it was written through and through with the assumption that `Unsigned n` wraps around. I never even realized it's not part of the intended interface. So then I run the Intel 8080 test suite, and it errors out in simulation with `"(+): overflow"`, but then finding where these additions and subtractions hide (in fun places like using `+=` from Lens) is very painful and only felt doable because the code is so small. 

Now, I realize existing code shouldn't be an argument against progress especially if the current behaviour is "wrong" or at least not intended to be exploited by user code, but the wraparound behaviour of `Unsigned` felt like such a natural thing that I'm sure I can't be the only one who ended up with code depending on it. 

(Option #2 sounds great by the way, I was never quite sure why `Index` doesn't wrap around; I ended up being vaguely convinced that it would be inefficient to do that).

BTW, I was so convinced that the wraparound behaviour of `Unsigned` is intended, that I even explicitly spelled it out it as such in the upcoming Retroclash book:

> [`Unsigned` and `Signed`] types are instances of `Eq` , `Ord` , `Num` etc. as one would expect (the `Num` instance is overflowing, e.g. `6 + 4 :: Unsigned 3` is `2`).

**Reply by alex-mckenna** [reply]
Additionally some of the overflow / underflow checking in the branch I linked above would also fail if erroring was the default

**Reply by martijnbastiaan** [reply]
I love this, thanks @gergoerdi. This is a pretty strong case against defaulting to error.

> Now, I realize existing code shouldn't be an argument against progress especially if the current behaviour is "wrong" or at least not intended to be exploited by user code,

I think it's a reasonable argument in cases where the compiler won't tell you to change your code after upgrading. 

> but then finding where these additions and subtractions hide (in fun places like using += from Lens) is very painful and only felt doable because the code is so small.

Tracking down bottoms is definitely too hard right now.

**Reply by rowanG077** [reply]
While I agree that migration is hard I feel this is also exactly the reason why you would want this change. The reason migration is hard is because it's unclear when an operation should be allowed to overflow or not so it's not a copy-paste level refactor. It's kind of a double edged sword either you get tortured during the refactor or your design might be subtly wrong because of silent overflows.

Another possibility for #2 would be to be fine with overflow behavior as it is currently and add an overflow error reporting newtype wrapper. 

Personally I would still prefer #3 with a reasonably long transition period so people have ample time to migrate but optional error reporting is kinda sexy too.

**Reply by martijnbastiaan** [reply]
> I was never quite sure why Index doesn't wrap around; I ended up being vaguely convinced that it would be inefficient to do that

@christiaanb Could you comment on this? Naively I would say that this is true, as I don't think there's a way without a `==` plus a mux to implement this.

Also, @christiaanb, a while ago you mentioned that stack traces _without_ profiling builds or `HasCallStack` constraints might be coming to GHC. If this has any chance of being merged, we might (and I think I prefer) want to do nothing for now (option 1) and reevaluate once that's in a sufficient number of GHCs supported by Clash.

**Reply by martijnbastiaan** [reply]
As an additional anecdote to this thread: I once spent two days tracking down the source of a bottom in a commercial codebase. It's no fun at all.

**Reply by christiaanb** [reply]
@martijnbastiaan  With regards to error vs wrap-around for `Index`: My mantra is basically: "performance above all". With Clash you build circuits, you are building circuits instead of writing software because you need that edge on performance. So the compiler and the standard library should do their best to give you the best performing circuit by default. If you want better functionality at the cost of performance, that should be an active choice. 

Adding wrap-around to the operators for `Index` results in an additional multiplexer (we could optimize this away for powers-of-two), thus increasing both area and propagation delay.

With regards to stack traces without profiling and without dwarf, see: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5456

**Comment by martijnbastiaan** [comment]
To everyone who replied on this thread:

* Making wrapping by default (option 2) is a no-go, as it would generate additional muxes over the status-quo. 
* Keeping everything as (option 1) or being loud about overflows everywhere (option 3) is still up for debate - with the added note that tracking down bottoms with today's Haskell/GHC is unreasonably hard.
* Thanks to @rowanG077's proposal and @alex-mckenna's implementation we now have a very nice way of switching wrapping modes on the type level (https://github.com/clash-lang/clash-compiler/tree/8a52fa339292a14e32bc865b3a4a456844991167/clash-prelude/src/Clash/Num).

Until we have "sane" stack trace support in GHC (https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5456), I'd like to close this discussion. Once that has landed we can revisit to see if option 3 would be reasonable. I'd like to close https://github.com/clash-lang/clash-compiler/pull/688 as this implements wrapping by default (option 2), which to reiterate has been rejected on (HDL) performance concerns.

**Reply by rowanG077** [reply]
@martijnbastiaan I agree, I have closed https://github.com/clash-lang/clash-compiler/pull/688 as it has been superseded by the wrappers.
